torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
       "Work needs to be completed before calling result(). "
       "Should call wait() before result().");
   TORCH_CHECK(
-      outputTensors_.size() <= 1, "work result does not support list of lists.");
+      outputTensors_.size() <= 1,
+      "work result does not support list of lists, use .getFuture() and value()");
   return outputTensors_.size() == 0 ? std::vector<at::Tensor>()
                                     : outputTensors_.at(0);
 }
 
+c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupGloo::AsyncWork::
+    getFuture() {
+  return future_;
+}
+
+namespace {
+c10::intrusive_ptr<c10::ivalue::Future> createFutureAsOutput(
+    const std::vector<std::vector<at::Tensor>>& outputTensors) {
+  if (outputTensors.size() > 1) {
+    return c10::make_intrusive<c10::ivalue::Future>(
+        c10::ListType::create(c10::ListType::create(c10::TensorType::get())));
+  }
+  return c10::make_intrusive<c10::ivalue::Future>(
+      c10::ListType::create(c10::TensorType::get()));
+}
+
+void returnFutureWithOutput(
+    c10::intrusive_ptr<c10::ivalue::Future> future,
```

Comment:
Nit: no need to pass the pointer by value, this function could just take a `Future&` and thus avoid the copy and the consequent refcount increase+decrease.